File: ios/FlowerLines/Game/GameState.swift
import Foundation

// MARK: - Constants
let GRID_SIZE   = 9
let NUM_TYPES   = 7
let LINE_MIN    = 5
let SPAWN_COUNT = 3

// MARK: - Game Phase
enum GamePhase {
    case idle, selected, animating, gameover
}

// MARK: - Grid Position
struct GridPos: Hashable, Equatable {
    let row: Int
    let col: Int
}

// MARK: - Game State
class GameState {
    var board: [[Int?]]          // nil = empty, 0-6 = flower type
    var score: Int
    var highScore: Int
    var selected: GridPos?
    var validMoves: Set<GridPos>?
    var nextFlowers: [Int]       // 3 upcoming types
    var animQueue: [AnimItem]
    var phase: GamePhase
    var pendingPhase: String?
    var eliminatingCells: [AnimCell]?
    var spawningCells: [AnimCell]?
    var animStart: Double        // timestamp ms

    init(highScore: Int = 0) {
        self.board = Array(repeating: Array(repeating: nil, count: GRID_SIZE), count: GRID_SIZE)
        self.score = 0
        self.highScore = highScore
        self.selected = nil
        self.validMoves = nil
        self.nextFlowers = []
        self.animQueue = []
        self.phase = .idle
        self.pendingPhase = nil
        self.eliminatingCells = nil
        self.spawningCells = nil
        self.animStart = 0
    }

    func emptyCells() -> [GridPos] {
        var cells: [GridPos] = []
        for r in 0..<GRID_SIZE {
            for c in 0..<GRID_SIZE {
                if board[r][c] == nil { cells.append(GridPos(row: r, col: c)) }
            }
        }
        return cells
    }
}

File: android/app/src/main/java/com/flowerlines/game/GameState.kt
package com.flowerlines.game

const val GRID_SIZE   = 9
const val NUM_TYPES   = 7
const val LINE_MIN    = 5
const val SPAWN_COUNT = 3
val PETAL_COUNTS = intArrayOf(5, 6, 8, 4, 6, 5, 7)

enum class GamePhase { IDLE, SELECTED, ANIMATING, GAMEOVER }

data class GridPos(val row: Int, val col: Int)

class GameState(var highScore: Int = 0) {
    // null = empty, 0-6 = flower type
    val board: Array<Array<Int?>> = Array(GRID_SIZE) { Array(GRID_SIZE) { null } }
    var score: Int = 0
    var selected: GridPos? = null
    var validMoves: Set<GridPos>? = null
    var nextFlowers: MutableList<Int> = mutableListOf()
    var animQueue: MutableList<AnimItem> = mutableListOf()
    var phase: GamePhase = GamePhase.IDLE
    var pendingPhase: String? = null
    var eliminatingCells: List<AnimCell>? = null
    var spawningCells: List<AnimCell>? = null
    var animStart: Long = 0L  // timestamp ms

    fun emptyCells(): List<GridPos> {
        val cells = mutableListOf<GridPos>()
        for (r in 0 until GRID_SIZE)
            for (c in 0 until GRID_SIZE)
                if (board[r][c] == null) cells.add(GridPos(r, c))
        return cells
    }

    fun resetBoard() {
        for (r in 0 until GRID_SIZE)
            for (c in 0 until GRID_SIZE)
                board[r][c] = null
    }
}
